### ISSUE TITLE
undefined reference to `INIT__0_sal__intgmp'

### ISSUE THREAD

**Issue by LdBeth** [pippijn/aldor#141]
When try to compile a program with `GMPInteger`, according to `libaldor` manual I use

```
$ ./aldor/bin/aldor -Fx -dGMP -cruntime=foam-gmp,gmp -laldor cow.as
/usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: cow.o: in function `INIT__0_cow':
cow.c:(.text+0x27c8): undefined reference to `INIT__0_sal__intgmp'
collect2: error: ld returned 1 exit status
#1 (Fatal Error) Linker failed.  Command was: /home/ldbeth/aldor/bin/unicl cow.o cow-aldormain.o -L. -L/home/ldbeth/aldor/share/lib -L/home/ldbeth/aldor/lib -o  cow -laldor -lfoam-gmp -lgmp
#1 (Warning) Removing file `cow.o'.
#2 (Warning) Removing file `cow-aldormain.o'.
```

The function `INIT__0_sal__intgmp` is in the C file generated from `aldor/aldor/lib/aldor/src/gmp/sal_intgmp.as`, however
using `nm` I didn't find it is defined In `libaldor.p`, and I didn't find the corresponding `*.o` in the building directory. Thus I suspect it is probably that the building script have been kept from compile and link this part of the library to the runtime.

**Comment by hemmecke** [comment]
As far as I remember, the libraries of Aldor must be compiled accordingly. You can create aldor-gmp.al and aldor-gmp.a as well as aldord.al and aldord.a. The first set with GMP support (-DGMP) and aldord was for debug support (-DDEBUG). I don't, however, currently recall how to compile these libraries.
In anyway, there are two Integer types GMPInteger and AldorInteger and Integer is just a macro that maps to either of them. Whereas in aldor.al it is AldorInteger in aldor-gmp.al it is GMPInteger.

**Comment by LdBeth** [comment]
Thank you for the info. It seems I need to add corresponding `aldor-gmp` entries and tweak the parameters to the `Makefile.am` under `lib/aldor/src` to have `make` build the gmp version of the library. Will try and test it out later.

**Comment by LdBeth** [comment]
```
--foo.as
#include "aldor"
#include "aldorio"

I ==> GMPInteger;

import from I;

x : I := 1999;

stdout << "This is: " << x << 1923 << newline;
```

I simply added `gmp/{sal_fltgmp.c,sal_gmptls.c,sal_intgmp.c}` to `Makefile.am` and the library builds without problem.

    aldor -Fx -dGMP -cruntime=foam-gmp,gmp -laldor foo.as

and with this command the above example file compiles.

However the output is ``This is: X```X`YZ``, probably there's something wrong in the printing function. I guess it should be as simple as tweaking some constants, but I don't have enough knowledge to fix it yet.

**Comment by LdBeth** [comment]
The definition of `<<` is

```
(p:TextWriter) << (x:%):TextWriter == {
                import from IntegerTypeTools %;
                print(x, 10@%, p);
        }
```

is `print` a builtin operation? Any idea on where the definition is?

**Comment by pbroadbery** [comment]
It's in IntegerTypeTools, arith/sal_itools.as.

The code looks to be a bit odd, but I can't quickly explain why you're
seeing that output.
I'll have a dig around as well.


Peter

**Comment by pbroadbery** [comment]
I have (I think) resolved this in #144 (branch is pab/gmp-doc-fixes) - Please take a look if you have time.
I need to do some further checks and then will merge unless any showstoppers show up.

**Comment by LdBeth** [comment]
Yes, I tested with `pab/gmp-doc-fixes` branch and the above small example can print  GMP numbers correctly.
